File: ios/Runner/FFmpegMethodChannel.swift
import Flutter
import ffmpegkit

class FFmpegMethodChannel : NSObject, FlutterPlugin, FlutterStreamHandler {
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
            
        case FFmpegMethodChannel.METHOD_EXECUTE:
            do {
                let arguments = call.arguments as! [String]
                
                let session = FFmpegSession.create(arguments)!
                FFmpegKitConfig.ffmpegExecute(session)
                
                var logs: String?
                if session.getReturnCode().isValueError() {
                    logs = session.getLogs().reversed().map { String(describing: $0) }.joined(separator: ", ")
                }
                let returnCode = FFmpegReturnCode(
                    isSuccess: session.getReturnCode().isValueSuccess(),
                    isCancel: session.getReturnCode().isValueCancel(),
                    isError: session.getReturnCode().isValueError(),
                    logs: logs
                )
                let json = String(data: try JSONEncoder().encode(returnCode), encoding: .utf8)!
                result(json)
            } catch {
                result(FlutterError(code: "ERROR", message: error.localizedDescription, details: error))
            }
            break
            
        case FFmpegMethodChannel.METHOD_DURATION:
            do {
                let path = call.arguments as! String
                let mediaInformation = FFprobeKit.getMediaInformation(path).getMediaInformation()!
                let duration = NumberFormatter().number(from: mediaInformation.getDuration())?.doubleValue ?? 0.0
                result(duration)
            } catch {
                result(FlutterError(code: "ERROR", message: error.localizedDescription, details: error))
            }
            break
            
        case FFmpegMethodChannel.METHOD_DISPOSE:
            FFmpegKit.cancel()
            result(true)
            break
            
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        let duration = (arguments as? Int).flatMap { max(Double($0), 0.1) } ?? 1.0
        FFmpegKitConfig.enableStatisticsCallback { statistic in
            let progress = min(max((statistic?.getTime() ?? 0.0) / duration, 0.0), 1.0)
            DispatchQueue.main.async { events(progress) }
        }
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        FFmpegKitConfig.enableStatisticsCallback {_ in }
        return nil
    }
    
    private static let NAME = "com.feelsoftware.slidemix.ffmpeg"
    private static let NAME_PROGRESS = "com.feelsoftware.slidemix.ffmpeg.progress"
    
    private static let METHOD_EXECUTE = "execute"
    private static let METHOD_DURATION = "duration"
    private static let METHOD_DISPOSE = "dispose"
    
    static func register(with registry: FlutterPluginRegistry) {
        let registrar = registry.registrar(forPlugin: FFmpegMethodChannel.NAME)
        if let safeRegistrar = registrar {
            register(with: safeRegistrar)
        }
    }
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let taskQueue = registrar.messenger().makeBackgroundTaskQueue!()
        let instance = FFmpegMethodChannel()
        registrar.addMethodCallDelegate(
            instance,
            channel: FlutterMethodChannel(
                name: FFmpegMethodChannel.NAME,
                binaryMessenger: registrar.messenger(),
                codec: FlutterStandardMethodCodec.sharedInstance(),
                taskQueue: taskQueue
            )
        )
        FlutterEventChannel(
            name: FFmpegMethodChannel.NAME_PROGRESS,
            binaryMessenger: registrar.messenger(),
            codec: FlutterStandardMethodCodec.sharedInstance(),
            taskQueue: taskQueue
        ).setStreamHandler(instance)
    }
    
}

private struct FFmpegReturnCode : Codable {
    let isSuccess : Bool
    let isCancel : Bool
    let isError : Bool
    let logs : String?
}

File: android/app/src/main/kotlin/com/feelsoftware/slidemix/ffmpeg/FFmpegMethodChannel.kt
package com.feelsoftware.slidemix.ffmpeg

import android.os.Handler
import android.os.Looper
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.FFmpegKitConfig
import com.arthenica.ffmpegkit.FFmpegSession
import com.arthenica.ffmpegkit.FFprobeKit
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMethodCodec
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

class FFmpegMethodChannel private constructor() :
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_EXECUTE -> {
                try {
                    val arguments = call.arguments<List<String>>()

                    val session = FFmpegSession.create(arguments.orEmpty().toTypedArray())
                    FFmpegKitConfig.ffmpegExecute(session)

                    val returnCode = FFmpegReturnCode(
                        isSuccess = session.returnCode.isValueSuccess,
                        isCancel = session.returnCode.isValueCancel,
                        isError = session.returnCode.isValueError,
                        logs = takeIf { session.returnCode.isValueError }
                            ?.let { session.logs.reversed().joinToString { it.message } }
                    )
                    val json = Json.encodeToString(returnCode)
                    result.success(json)
                } catch (error: Throwable) {
                    result.error("ERROR", error.message, error.stackTraceToString())
                }
            }

            METHOD_DURATION -> {
                try {
                    val path = call.arguments<String>()!!
                    val mediaInformation = FFprobeKit.getMediaInformation(path).mediaInformation
                    val duration = mediaInformation.duration.toDoubleOrNull() ?: 0.0
                    result.success(duration)
                } catch (error: Throwable) {
                    result.error("ERROR", error.message, error.stackTraceToString())
                }
            }

            METHOD_DISPOSE -> {
                FFmpegKit.cancel()
                result.success(true)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        val duration = (arguments as? Int)?.toDouble()?.coerceAtLeast(0.1) ?: 1.0
        val handler = Handler(Looper.getMainLooper())
        FFmpegKitConfig.enableStatisticsCallback { statistic ->
            val progress = (statistic.time / duration)
                .coerceAtLeast(0.0)
                .coerceAtMost(1.0)
            handler.post { events.success(progress) }
        }
    }

    override fun onCancel(arguments: Any?) {
        FFmpegKitConfig.enableStatisticsCallback {}
    }

    companion object {

        private const val NAME = "com.feelsoftware.slidemix.ffmpeg"
        private const val NAME_PROGRESS = "com.feelsoftware.slidemix.ffmpeg.progress"

        private const val METHOD_EXECUTE = "execute"
        private const val METHOD_DURATION = "duration"
        private const val METHOD_DISPOSE = "dispose"

        fun attach(flutterEngine: FlutterEngine) {
            val implementation = FFmpegMethodChannel()
            val taskQueue = flutterEngine.dartExecutor.binaryMessenger.makeBackgroundTaskQueue()
            MethodChannel(
                flutterEngine.dartExecutor.binaryMessenger,
                NAME,
                StandardMethodCodec.INSTANCE,
                taskQueue,
            ).setMethodCallHandler(implementation)
            EventChannel(
                flutterEngine.dartExecutor.binaryMessenger,
                NAME_PROGRESS,
                StandardMethodCodec.INSTANCE,
                taskQueue,
            ).setStreamHandler(implementation)
        }
    }
}

@Serializable
private data class FFmpegReturnCode(
    val isSuccess: Boolean,
    val isCancel: Boolean,
    val isError: Boolean,
    val logs: String?,
)
